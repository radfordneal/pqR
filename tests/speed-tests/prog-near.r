source("time.r")
 
test.name <- "near"

# Find squared distances of columns of M from vector v.

sq_dist <- test.cmp (function (M, v) colSums((M-v)^2))

# Find a column of M that is closest to each column of V, from among those 
# columns whose projections by P are at squared distance no more than t from
# the projection by P of V.  Returns a list with elements "count", containing
# the number of columns of M within projected distance t for each column 
# of V, and "dist", containing the distance of the column found from V
# (Inf if no column found).  The projection of M by P may be passed as
# a pre-computed argument.

near <- test.cmp (function (M,V,P,t,proj_M=P%*%M)
{ 
  count <- numeric(ncol(V))
  dist <- numeric(ncol(V))

  for (j in 1:ncol(V))
  { 
    sqd <- sq_dist (proj_M, as.vector(P %*% V[,j]))

    c <- 0
    d <- Inf

    for (i in 1:length(sqd))
    { if (sqd[i] < t)
      { c <- c + 1
        d2 <- sum((M[,i]-V[,j])^2)
        if (d2 < d) d <- d2
      }
    }

    count[j] <- c
    dist[j] <- sqrt(d)
  }

  list (count=count, dist=dist)
})

set.seed(1)

P <- matrix(runif(100,-2,+2),5,20)
M <- matrix(rnorm(4000),20,200)
proj_M <- P %*% M

V <- matrix(rnorm(400),20,20)

# Note: The call of unlist below should force completion of all computations.

sys.time (for (i in 1:500) res <- unlist(near(M,V,P,8^2,proj_M)))

print(res)

set.seed(2)

P <- matrix(runif(100000,-2,+2),100,1000)
M <- matrix(rnorm(20000000),1000,20000)
proj_M <- P %*% M

V <- matrix(rnorm(500000),1000,500)

sys.time (n <- near(M,V,P,425^2,proj_M))

print(c(mean(n$count),mean(n$count==0)))
print(n$dist[c(1,500)])
