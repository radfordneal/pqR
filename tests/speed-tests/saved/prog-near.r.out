
---- near 

Time:
   count1    count2    count3    count4    count5    count6    count7    count8 
 2.000000 22.000000 21.000000  7.000000  3.000000 11.000000 21.000000 11.000000 
   count9   count10   count11   count12   count13   count14   count15   count16 
31.000000  5.000000 19.000000 15.000000 18.000000  1.000000 24.000000 22.000000 
  count17   count18   count19   count20     dist1     dist2     dist3     dist4 
15.000000 18.000000 26.000000 26.000000  5.253719  3.604315  3.580123  5.638031 
    dist5     dist6     dist7     dist8     dist9    dist10    dist11    dist12 
 6.112476  4.632962  4.093481  3.833922  4.808695  4.844333  5.508277  4.076000 
   dist13    dist14    dist15    dist16    dist17    dist18    dist19    dist20 
 4.659942  5.713242  3.722731  4.137022  4.382561  3.753332  3.651141  3.808293 
Time:
[1] 118.852   0.044
[1] 42.27782 42.03465
